feat: Add unique id for every memory consumer

EmilyMatt opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

- Closes #15126 .

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

This makes an important change to the MemoryConsumer API, letting any memory pool identify which consumer is requesting the memory, and therefor separate consumers that might have the same name (for example, two AggregateExec operators in the same Partition)
This also modifies the TrackConsumer pool to adhere to these changes.

## What changes are included in this PR?
Nothing except the addition of  a unique identifier in the consumer, and an id() function to retrieve it.

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

All of the TrackedConsumer pool tests now verify this, as they have all been updated to accomodate the fact that the pool uses the IDs to track (the inner pool is unchanged in its behaviour atm)
This ensures the validity of the logic

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

Yes, MemoryConsumer is no longer clone-able, it is now moved into the register() function and thats it, to avoid an invalid state where a MemoryConsumer can be registered more than once.
The TrackedConsumerMemoryPool will now store Consumers based on their IDs, instead of names, so will be able to track multiple consumers with the same name, based on their ID, and the error messages are updated accordingly.

alamb reviewed (commented):
Thank you @EmilyMatt  -- this API make a lot of sense to me

I think it would be useful to add tests for this feature, so that we don't break it in the future accidentally

Specifically, I think testing the "memory consumer ids are unique" behavior would be best as that is the goal of the API. SO for example, perhaps write a query that has multiple `AggregateExec`s and verify that the assigned memory consumer Ids are different

I would also recommend adding ids to the various messages (such as in TrackedConsumerPool) https://github.com/apache/datafusion/blob/main/datafusion/execution/src/memory_pool/pool.rs#L261

So that the id can be correlated in more places

EmilyMatt commented:
> Thank you @EmilyMatt -- this API make a lot of sense to me
> 
> I think it would be useful to add tests for this feature, so that we don't break it in the future accidentally
> 
> Specifically, I think testing the "memory consumer ids are unique" behavior would be best as that is the goal of the API. SO for example, perhaps write a query that has multiple `AggregateExec`s and verify that the assigned memory consumer Ids are different
> 
> I would also recommend adding ids to the various messages (such as in TrackedConsumerPool) https://github.com/apache/datafusion/blob/main/datafusion/execution/src/memory_pool/pool.rs#L261
> 
> So that the id can be correlated in more places

I have modified this PR to also reflect these changes in the TrackConsumer pool, but in order to ensure a valid state this also requires a breaking API change - MemoryConsumer no longer implements Hash and Clone
This is in order to ensure a user does not call clone() on a consumer, and then register the cloned one, while still maintaining an owned MemoryConsumer, which can be called with register() again, despite having the same id as an already existing MemoryConsumer in the pool.

I'd have preferred to deprecate the Clone implementation and only removing it in a future version, along with an explanation, but unfortunately, traits cannot be deprecated using the attribute.
This means this PR needs the api change label, but I can't manually set it^

alamb commented on datafusion/execution/src/memory_pool/mod.rs:
```diff
@@ -149,21 +149,48 @@ pub trait MemoryPool: Send + Sync + std::fmt::Debug {
 ///
 /// [proxy]: datafusion_common::utils::proxy
-#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+#[derive(Debug)]
```
It is still ok for this struct to derive Hash, right?

In terms of clone, perhaps we can add a method like `pub fn clone_with_new_id(&self)` that can be explicitly called to clone the memory consumer while providing a new id

EmilyMatt replied:
Well a Derive can't be used since we don't derive the PartialEq and Eq, but I will add a full Hash function that will take into consideration the id as well I think, I think a straightforward change is still what I would prefer from a crate I'm using

berkaysynnada reviewed (commented):
Are we planning to deprecate MemoryConsumer and replace it all with TrackedConsumer ?

EmilyMatt commented:
> Are we planning to deprecate MemoryConsumer and replace it all with TrackedConsumer ?

I think TrackedConsumer is only relevant for the TrackConsumerPool, it doesn't exist anywhere else, most pools don't track consumers at all, it is up to users to implement a MemoryPool that takes into consideration each Consumer and how much it has requested.
This API is only meant to help users do that, as the current API does not really allow doing that with certainty(there can be surrounding hacks to achieve that functionality, but they far from perfect without some unique identifier)

alamb commented on datafusion/execution/src/memory_pool/mod.rs:
```diff
@@ -149,21 +150,65 @@ pub trait MemoryPool: Send + Sync + std::fmt::Debug {
+        }
+    }
+
+    pub fn clone_with_new_id(&self) -> Self {
```
```suggestion
    /// Return a clone of the MemoryConsumer that has a new
    /// unique ID assigned. 
    pub fn clone_with_new_id(&self) -> Self {
```

EmilyMatt replied:
No problem, added some more documentation in general

alamb commented on datafusion/execution/src/memory_pool/mod.rs:
```diff
@@ -149,21 +150,65 @@ pub trait MemoryPool: Send + Sync + std::fmt::Debug {
 ///
 /// [proxy]: datafusion_common::utils::proxy
-#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+#[derive(Debug)]
```
Can you also please add a note to this doc string that says each MemoryConsumer gets an automatically assigned globally unique id and that to clone a memory consumer, see [`MemoryConsumer::clone_with_new_id`]

EmilyMatt replied:
Addressed^

alamb commented on datafusion/execution/src/memory_pool/pool.rs:
```diff
@@ -261,7 +268,7 @@ fn insufficient_capacity_err(
     inner: I,
     top: NonZeroUsize,
-    tracked_consumers: Mutex<HashMap<MemoryConsumer, AtomicU64>>,
+    tracked_consumers: Mutex<HashMap<usize, TrackedConsumer>>,
```
```suggestion
    /// Maps consumer_id --> TrackedConsumer
    tracked_consumers: Mutex<HashMap<usize, TrackedConsumer>>,
```

EmilyMatt replied:
Added a documentation to all the fields there^

alamb commented on datafusion/execution/src/memory_pool/pool.rs:
```diff
@@ -277,40 +284,29 @@ impl<I: MemoryPool> TrackConsumersPool<I> {
+                        tracked_consumer.name.to_owned(),
+                        tracked_consumer.can_spill,
+                    ),
+                    tracked_consumer.reserved.load(Ordering::Acquire),
```
As a follow on / cleanup it would be possible to refactor this code into methods, so like this (and it might make the code easier to read)

```suggestion
                    tracked_consumer.reserved(),
```

alamb replied:
The same thing goes for other updates later down

EmilyMatt replied:
Sounds good to me, done^

alamb commented on datafusion/execution/src/memory_pool/pool.rs:
```diff
@@ -501,12 +507,12 @@ mod tests {
         // using the previously set sizes for other consumers
         let mut r5 = MemoryConsumer::new("r5").register(&pool);
-        let expected = "Additional allocation failed with top memory consumers (across reservations) as: r1 consumed 50 bytes, r3 consumed 20 bytes, r2 consumed 15 bytes. Error: Failed to allocate additional 150 bytes for r5 with 0 bytes already allocated for this reservation - 5 bytes remain available for the total pool";
+        let expected = format!("Additional allocation failed with top memory consumers (across reservations) as: r1#{}(can spill: false) consumed 50 bytes, r3#{}(can spill: false) consumed 20 bytes, r2#{}(can spill: false) consumed 15 bytes. Error: Failed to allocate additional 150 bytes for r5 with 0 bytes already allocated for this reservation - 5 bytes remain available for the total pool", r1. ...
```
this is a nice change

alamb reviewed (approved):
Thank you @EmilyMatt  -- I think this looks like a very nice improvement to me

FYI @wiedld  and @2010YOUY01 

I also marked this PR as an API change as MemoryConsumer no longer implements Clone.

EmilyMatt commented:
@alamb Thanks for taking the time to look at this. I've addressed the relevant points and added some general documentation, I've also modified the reserved field to an AtomicUsize(instead of AtomicU64), as that feels more inline with the memory pool API, which uses usize everywhere, I hope that's not overtly obtrusive.

alamb commented:
This is great -- thanks again for the work and contribution @EmilyMatt
